greenriver/hmis-warehouse
# Use a job to enforce queing discipline for WarehouseClientsProcessed

## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch

## Description
[//]: # (Summarize changes and include links related issue)
[//]: # (List any new dependencies or relevant ADRs)
* Replaces `update_cached_counts_no_named_arguments` in `GrdaWarehouse::WarehouseClientsProcessed` with a full-fledged Job class.
* Enforces sequential-ish generating of cached warehouse client data by checking for a running job and re-queuing for a few minutes later if there is one 

To test:
1. Run: `GrdaWarehouse::WarehouseClientsProcessed.update_cached_counts` (where you have more than 5,000 destination clients in the database
2. Fire up 2 or more delayed job workers and watch the DJ queue. 
3. You should see a bunch of jobs created in step 1
4. You should see one job run and the remaining re-queued for a few minutes later, the process should repeat until no jobs remain

## Type of change
[//]: # (e.g., Bug fix, New feature, Documentation, Code clean-up, Dependency update)
* Code clean-up

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)

[//]: # NOTE: system tests may fail if there is no branch on the hmis-frontend that matches the Source  or Target branch of this PR. This is expected
